pandas/core/common.py

```diff
@@ -463,8 +466,31 @@ def random_state(state: RandomState | None = None):
         )
 
 
+@overload
+def pipe(
+    obj: S,
```

Comment:
Good catch. I'll look into this later.

Reply:
Ok I think I changed my mind on this. Nothing in `pipe`'s implimentation is specific to the type `NDFrame`, so it seem more accurate to use a generic type var (`S` in this case).

I think changing it to `NDFrameT` gives some errors, as it's used by `BaseGroupType` and `Styler`—which are not subclasses of `NDFrame`.

Reply:
Okay makes sense!

I believe we try to avoid one-letter variables and replace them with more descriptive variable names. I know `T` and `P` are common for a plain `TypeVar` and `ParamSpec`. Can you replace `S` with `T1` (or something similar)?

Reply:
>  I know `T` and `P` are common for a plain `TypeVar` and `ParamSpec`.

Similarly `S` is also a common naming-convention used name for your second `TypeVar` if `T` already exists, like it does here (I think!). Some examples in PEP 484:

https://peps.python.org/pep-0484/#:~:text=A%20generic%20type,S%5D)%3A%0A%20%20%20%20...

I'm not married to this though, I'm ok with `T1` as well. More of an FYI. I'll leave it up to you!

Reply:
I have no strong opinion, but I assume `S` (and `P`) will not be used very often inside pandas, meaning people might not have a mental model of what they are. `T` isn't ideal either, but it is more common in the pandas (and everywhere in the typing community).

@Dr-Irv @jbrockmendel (I believe you opened an issue at some point about too short names)

Reply:
There's no issue, just a general preference to avoid 1-letter names, largely for grep-ability purposes.  if no one else cares, don't worry about me on this one.

Reply:
Ok, seeming as everyone here is pretty ambivalent about this one, I'm going to resolve this andf keep it as `S`—for no better reason other than it being the current typing naming convention, which we will follow in the absence of any opinion.

Again if anyone does come along with a strong opinion on this, happy to rename!